### PR TITLE
Swapped to using Hydra dataclasses

### PR DESCRIPTION
**Pull request by SeanNaren**
Attempt at #545 

@omry would I be able to get you to have a look at my first stab at adding data classes here? I'd also like to move the conf into its own file; would that be fine?

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -1,11 +1,107 @@
 import hydra
+from attr import dataclass
```
Comment (omry) [inline train.py:2]: that's odd.
If you want to use attr, you can use it directly (OmegaConf supports attr classes).
If you want to use dataclass, you should import it from dataclasses (3.7+, there is a dataclasses pip package for 3.6).
```
from dataclasses import dataclass
```

File: train.py [hunk 2]
```diff
@@ -1,11 +1,107 @@
...
 
 
-@hydra.main(config_path="config", strict=False)
-def hydra_main(cfg: DictConfig):
+@dataclass
+class TrainConfig:
+    no_cuda: bool = False  # Enable CPU only training
+    finetune: bool = False  # Fine-tune the model from checkpoint "continue_from"
+    seed: int = 123456  # Seed for generators
+    dist_backend: str = 'nccl'  # If using distribution, the backend to be used
```
Comment (omry) [inline train.py:14]: if you have a closed set choices for dist_backend, use an Enum.

**Comment by SeanNaren** [comment]
> This is not config groups, just dataclasses creating a single config object.
> 
> Go through the basic tutorial to understand config groups.

Thanks omry! I had a feeling this wasn't taking advantage of config groups but i'm a bit unsure why it would be needed/preferred?

How would grouping of configs help further than what I've implemented? It seems like it would be useful when specifying different configurations but given that there is only one configuration supported it doesn't directly complete the functionality here unless I'm wrong.

What I guess would make sense to implement would be various configs for different functionality that can then be built together to make the run (CPU only config/GPU config with other smaller configs to combine)

**Comment by omry** [comment]
It all depends on what you want to do.
If at this point you have a static structure for your config what you did is perfectly fine.

config groups allow you to compose your config dynamically. 
As an example, you have OptimizerConfig which some configs for your optimizer.
What if you want to support both Adam and Nesterov? each one got their own config options. you don't really want to squash them together.
Config groups allow you to choose one or the other and have a clean ad minimal config for each mode.

This also allow you to sweep over `optimizer=adam,nesterov`, where each of the choices will bring in a bunch of default config values.
essentially, it lets you get to a higher level of abstraction.

**Comment by SeanNaren** [comment]
Cheers, I'll get some time to experiment with the idea, I think there is potential here for config groups to be useful!

**Comment by SeanNaren** [comment]
Made a few changes and added a config group for the optimizer. I think other parameters unfortunately do not fit the schema yet, but maybe in the future :)

The configs are slowly creeping their way into the rest of the code which is nice, however due to backwards compat and silly decisions made in the past I need to be careful how much gets changed.

I'll run various configs to see how things work, but then shortly after I'll be working on getting tests integrated into this repo.

File: README.md [hunk 3]
```diff
@@ -87,22 +87,23 @@ python merge_manifests.py --output-path merged_manifest.csv --merge-dir all-mani
...
+Configuration is done via [Hydra](https://github.com/facebookresearch/hydra).
+
+Defaults can be seen in [config.py](./deepspeech_pytorch/config.py). Below is how you can override values set already:
 
 ```
 python train.py data.train_manifest=data/train_manifest.csv data.val_manifest=data/val_manifest.csv
 ```
 
-Use `python train.py --help` for more parameters and options.
+Use `python train.py --help` for all parameters and options.
```
Comment (omry) [inline README.md:98]: By the way, you can customize your help a bit:
https://hydra.cc/docs/configure_hydra/app_help

Check the examples in Hydra for a complete example (I need to update the doc to link to it).

File: deepspeech_pytorch/config.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+
+@dataclass
+class ApexConfig:
+    opt_level: str = 'O1'  # Apex optimization level, check https://nvidia.github.io/apex/amp.html for more information
+    loss_scale: int = 1  # Loss scaling used by Apex. Default is 1 due to warp-ctc not supporting scaling of gradients
+
+
+@dataclass
+class DeepSpeechConfig(DictConfig):
```
Comment (omry) [inline deepspeech_pytorch/config.py:105]: I am going to eliminate the need to extend DictConfig just to get type safe access to cfg.pretty() before Hydra 1.0 is out.

The change change is likely to be to deprecated pretty() in favor of OmegaConf.pretty().
I might also change the default of resolve to True in the process.

File: deepspeech_pytorch/config.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    {"optim": "sgd"}
+]
+
+
+@dataclass
+class TrainingConfig:
+    no_cuda: bool = False  # Enable CPU only training
+    finetune: bool = False  # Fine-tune the model from checkpoint "continue_from"
+    seed: int = 123456  # Seed for generators
+    dist_backend: DistributedBackend = DistributedBackend.nccl  # If using distribution, the backend to be used
```
Comment (omry) [inline deepspeech_pytorch/config.py:18]: Awesome, did you try to make a command line override error here yet? :)

File: README.md (hunk 3 above)
Comment (SeanNaren) [reply README.md:98]: Awesome! any chance helper message per argument is on the roadmap? I guess it's not the cleanest thing to implement though

File: deepspeech_pytorch/config.py (hunk 5 above)
Comment (SeanNaren) [reply deepspeech_pytorch/config.py:18]: Seems to work great!!

File: README.md (hunk 3 above)
Comment (omry) [reply README.md:98]: I am going to try something better eventually.

https://github.com/omry/omegaconf/issues/131

**Comment by SeanNaren** [comment]
The RNN type is annoying me but I'll make an issue to track the deprecation of the old pattern once a new model release comes around!

**Comment by omry** [comment]
Model is a good candidate for a config group.

**Comment by SeanNaren** [comment]
Good shout, @omry is it possible to allow for an optional a config file? The use case would be if someone has changed a lot of commands via overrides and would like to create a custom yaml file with their changes.

**Comment by omry** [comment]
Yes. create a config group, say experiment:
```
experiment/
  foo.yaml
```

And add it from the command line:
```
python foo.py +experiment=foo
```

If you need a different defaults list there, you will have to use --config-file (which should be fully specified).

**Comment by SeanNaren** [comment]
I think i'm happy with the integration here! Going forward I'll probably make a release branch with breaking changes to get rid of some of these high level if statements and start to move the configs through the rest of the internals.

Also during the release will define some config files to represent each dataset and hyper-parameters, much neater :)
